### PR TITLE
Np 44962 fetch person preferences

### PR DESCRIPTION
**Pull request by EyobSikt**

### CODE REVIEW

File: preferences-commons/src/main/java/no/sikt/nva/person/preferences/commons/model/PersonPreferences.java [hunk 1]
```diff
@@ -13,7 +14,8 @@
 @JsonSerialize
 public record PersonPreferences(URI personId,
                                 List<URI> promotedPublications,
-                                @JsonIgnore PersonPreferencesService personPreferencesService) {
+                                @JsonIgnore PersonPreferencesService personPreferencesService,
```
Comment (truhacevkir) [inline preferences-commons/src/main/java/no/sikt/nva/person/preferences/commons/model/PersonPreferences.java:17]: We do not want to add created and modified dates to PersonPreferences, but only Dao

File: preferences-commons/src/main/java/no/sikt/nva/person/preferences/commons/service/PersonPreferencesService.java [hunk 2]
```diff
@@ -48,6 +48,8 @@ public PersonPreferences getPreferencesByPersonId(URI personId) {
         return new PersonPreferences.Builder()
                    .withPersonId(dao.personId())
                    .withPromotedPublications(dao.promotedPublications())
+                   .withCreated(dao.created())
```
Comment (truhacevkir) [inline preferences-commons/src/main/java/no/sikt/nva/person/preferences/commons/service/PersonPreferencesService.java:51]: Dao only should have those values

File: preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    public FetchPersonPreferencesHandler(PersonPreferencesService personPreferencesService) {
+        super(Void.class);
+        this.personPreferencesService = personPreferencesService;
+    }
+
+    @Override
+    protected PersonPreferences processInput(Void input, RequestInfo requestInfo, Context context)
+        throws ApiGatewayException {
+
+        statusCode = HttpURLConnection.HTTP_OK;
```
Comment (truhacevkir) [inline preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java:39]: Do we need to do this?

File: preferences-rest/src/test/java/no/sikt/nva/person/preferences/rest/UpdatePreferencesHandlerTest.java [hunk 4]
```diff
@@ -49,8 +49,8 @@ void shouldUpdatePersonPreferences() throws IOException {
                                           .build();
         handler.handleRequest(createRequest(updatePersonPreferences), output, CONTEXT);
 
-        assertThat(personPreferencesService.getPreferencesByPersonId(existingPersonPreferences.personId()),
-                   is(equalTo(updatePersonPreferences)));
+        assertThat(personPreferencesService.getPreferencesByPersonId(existingPersonPreferences.personId()).personId(),
```
Comment (truhacevkir) [inline preferences-rest/src/test/java/no/sikt/nva/person/preferences/rest/UpdatePreferencesHandlerTest.java:52]: Why?

File: template.yaml [hunk 5]
```diff
@@ -192,4 +192,27 @@ Resources:
...
+        ApplyOn: PublishedVersions
+      Environment:
+        Variables:
+          ALLOWED_ORIGIN: '*'
+          TABLE_NAME: !Ref NvaPersonPreferencesTable2
+      Events:
+        GetEvent:
+          Type: Api
+          Properties:
+            Path: /{personId+}
```
Comment (truhacevkir) [inline template.yaml:216]: Do we need `+` at the end of path parameter?

File: docs/openapi.yaml [hunk 6]
```diff
@@ -80,6 +80,30 @@ paths:
             application/json:
               schema:
                 $ref: "#/components/schemas/Problem"
+  /{personId+}:
+    parameters:
+      - in: path
+        name: personId+
```
Comment (truhacevkir) [inline docs/openapi.yaml:86]: Do we need `+` at the end of path parameter?

File: preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+        var personId = getPersonId(requestInfo);
+        logger.info("requestInfo.personId: " + personId);
+        return personPreferencesService.getPreferencesByPersonId(personId);
+    }
```
Comment (truhacevkir) [inline preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java:44]: I think we want to throw exception when there is no entry with given id.

File: preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    protected PersonPreferences processInput(Void input, RequestInfo requestInfo, Context context)
+        throws ApiGatewayException {
+
+        statusCode = HttpURLConnection.HTTP_OK;
+
+        var personId = getPersonId(requestInfo);
+        logger.info("requestInfo.personId: " + personId);
```
Comment (truhacevkir) [inline preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java:42]: No need to log id

**Comment by truhacevkir** [comment]
What about creating method fetch() on PersonPreferences as we do when updating and creating. 

For example:

```
        new PersonPreferences.Builder(personPreferencesService)
            .withPersonId(requestInfo.getPersonCristinId())
            .build()
            .fetch();
```

File: preferences-commons/src/main/java/no/sikt/nva/person/preferences/commons/model/PersonPreferences.java (hunk 1 above)
Comment (EyobSikt) [reply preferences-commons/src/main/java/no/sikt/nva/person/preferences/commons/model/PersonPreferences.java]: Removed.

File: preferences-commons/src/main/java/no/sikt/nva/person/preferences/commons/service/PersonPreferencesService.java (hunk 2 above)
Comment (EyobSikt) [reply preferences-commons/src/main/java/no/sikt/nva/person/preferences/commons/service/PersonPreferencesService.java]: Done.

File: preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java (hunk 3 above)
Comment (EyobSikt) [reply preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java]: Fixed.

File: preferences-rest/src/test/java/no/sikt/nva/person/preferences/rest/UpdatePreferencesHandlerTest.java (hunk 4 above)
Comment (EyobSikt) [reply preferences-rest/src/test/java/no/sikt/nva/person/preferences/rest/UpdatePreferencesHandlerTest.java:52]: Fixed. Removed ..personId().

File: preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java (hunk 8 above)
Comment (EyobSikt) [reply preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java]: Done.

File: preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java (hunk 7 above)
Comment (EyobSikt) [reply preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java:48]: Done.

File: preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    protected PersonPreferencesDao processInput(Void input, RequestInfo requestInfo, Context context)
+        throws ApiGatewayException {
+
+        var personId = getPersonId(requestInfo);
+        return attempt(() -> personPreferencesService.fetchPreferencesByPersonId(personId))
+            .orElseThrow(personPreferencesNotFound(personId));
+    }
```
Comment (truhacevkir) [inline preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java:43]: Object returned from API should not be DAO, in this case you should return PresonPreferences

File: preferences-rest/src/test/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+        handler.handleRequest(request, output, CONTEXT);
+        var response = GatewayResponse.fromOutputStream(output, PersonPreferencesDao.class);
+
+        assertThat(personPreferencesService.fetchPreferencesByPersonId(personPreferences.personId()),
+                   is(equalTo(response.getBodyObject(PersonPreferencesDao.class))));
+    }
+
+    @Test
+    void shouldReturnNotFoundWhenPersonIdentifierDoesNotExist() throws IOException {
```
Comment (truhacevkir) [inline preferences-rest/src/test/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandlerTest.java:62]: nice

File: preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java (hunk 9 above)
Comment (truhacevkir) [inline preferences-rest/src/main/java/no/sikt/nva/person/preferences/rest/FetchPersonPreferencesHandler.java:43]: One more thing, you do not need to send personId as pathParameter, cause you have this id in token already.
Use this method to access cristinId:
`requestInfo.getPersonCristinId()`

You should also change openapi and template scheme :)

**Review by truhacevkir — APPROVED** [review]

**Pull request merged**